src/cascadia/TerminalSettingsEditor/ActionsViewModel.cpp

```diff
@@ -18,383 +24,1194 @@ using namespace winrt::Windows::UI::Xaml::Data;
+            ModifyKeyChordRequested.raise(*this, *args);
+        }
+        else
+        {
+            // no changes being requested, toggle edit mode
+            ToggleEditMode();
+        }
+    }
+
+    void KeyChordViewModel::CancelChanges()
+    {
+        ToggleEditMode();
+    }
+
+    void KeyChordViewModel::DeleteKeyChord()
+    {
+        DeleteKeyChordRequested.raise(*this, _currentKeys);
+    }
+
+    ActionsViewModel::ActionsViewModel(Model::CascadiaSettings settings) :
+        _Settings{ settings }
+    {
+        _MakeCommandVMsHelper();
+    }
+
+    void ActionsViewModel::UpdateSettings(const Model::CascadiaSettings& settings)
+    {
+        _Settings = settings;
+
+        // We want to re-initialize our CommandList, but we want to make sure
+        // we still have the same CurrentCommand as before (if that command still exists)
+
+        // Store the ID of the current command
+        const auto currentCommandID = CurrentCommand() ? CurrentCommand().ID() : hstring{};
+
+        // Re-initialize the command vm list
+        _MakeCommandVMsHelper();
+
+        // Re-select the previously selected command if it exists
+        if (!currentCommandID.empty())
+        {
+            const auto it = _CommandList.First();
+            while (it.HasCurrent())
             {
-                const auto newFocusedIndex{ std::clamp(index, 0u, _KeyBindingList.Size() - 1) };
-                // Emit an event to let the page know to move focus to
-                // this VM's container.
-                FocusContainer.raise(*this, winrt::box_value(newFocusedIndex));
+                auto cmd = *it;
+                if (cmd.ID() == currentCommandID)
+                {
+                    CurrentCommand(cmd);
+                    break;
+                }
+                it.MoveNext();
+            }
+            if (!it.HasCurrent())
+            {
+                // we didn't find the previously selected command
+                CurrentCommand(nullptr);
+                CurrentPage(ActionsSubPage::Base);
             }
         }
+        else
+        {
+            // didn't have a command,
+            // so skip over looking through the command
+            CurrentCommand(nullptr);
+            CurrentPage(ActionsSubPage::Base);
+        }
     }
 
-    void ActionsViewModel::_KeyBindingViewModelModifyKeyBindingHandler(const Editor::KeyBindingViewModel& senderVM, const Editor::ModifyKeyBindingEventArgs& args)
+    void ActionsViewModel::_MakeCommandVMsHelper()
     {
-        const auto isNewAction{ !args.OldKeys() && args.OldActionName().empty() };
+        // Populate AvailableActionsAndNames
+        _AvailableActionsAndNamesMap = Model::CascadiaSettings::AvailableShortcutActionsAndNames();
+        for (const auto unimplemented : UnimplementedShortcutActions)
+        {
+            _AvailableActionsAndNamesMap.Remove(unimplemented);
+        }

```

Comment:
We can probably cache this, right? Looks like `CascadiaSettings::AvailableShortcutActionsAndNames` does the work each time we call it, then we'll remove the same 2 unimplemented shortcut actions. Maybe we should just store it as a static so we don't regenerate it? It shouldn't change at runtime.